hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclController.java

```diff
@@ -0,0 +1,667 @@
+          pathHelper.getArchiveNsDir(ns.getName()), pathHelper.getMobDataNsDir(ns.getName()) };
+      for (Path nsDir : nsDirs) {
+        if (!fs.exists(nsDir)) {
+          fs.mkdirs(nsDir);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void postCompletedSnapshotAction(ObserverContext<MasterCoprocessorEnvironment> c,
+      SnapshotDescription snapshot, TableDescriptor tableDescriptor) throws IOException {
+    if (!tableDescriptor.getTableName().isSystemTable() && checkInitialized()) {
+      hdfsAclHelper.snapshotAcl(snapshot);
+    }
+  }
+
+  @Override
+  public void postCompletedTruncateTableAction(ObserverContext<MasterCoprocessorEnvironment> c,
+      TableName tableName) throws IOException {
+    if (!tableName.isSystemTable() && checkInitialized()) {
+      hdfsAclHelper.resetTableAcl(tableName);
+    }
+  }
+
+  @Override
+  public void postDeleteTable(ObserverContext<MasterCoprocessorEnvironment> ctx,
+      TableName tableName) throws IOException {
+    if (!tableName.isSystemTable() && checkInitialized()) {
+      // remove table user HDFS acl from ns data directory if the user has no global/the ns/other
+      // tables of the ns permissions
+      Set<String> removeUsers = new HashSet<>();
+      try (Table aclTable =
+          ctx.getEnvironment().getConnection().getTable(PermissionStorage.ACL_TABLE_NAME)) {
+        List<String> users = SnapshotScannerHDFSAclStorage.getTableUsers(aclTable, tableName);
+        SnapshotScannerHDFSAclStorage.deleteTableHdfsAcl(aclTable, tableName);
+        for (String user : users) {
+          List<byte[]> userEntries = SnapshotScannerHDFSAclStorage.getUserEntries(aclTable, user);
+          boolean remove = true;
+          for (byte[] entry : userEntries) {
+            if (PermissionStorage.isGlobalEntry(entry)) {
+              remove = false;
+              break;
+            } else if (PermissionStorage.isNamespaceEntry(entry) && Bytes
+                .equals(PermissionStorage.fromNamespaceEntry(entry), tableName.getNamespace())) {
+              remove = false;
+              break;
+            } else if (Bytes.equals(TableName.valueOf(entry).getNamespace(),
+              tableName.getNamespace())) {
+              remove = false;
+              break;
+            }
+          }
+          if (remove) {
+            removeUsers.add(user);
+          }
+        }
+      }
+      if (removeUsers.size() > 0) {
+        hdfsAclHelper.removeNamespaceAcl(tableName, removeUsers);
+      }
+    }
+  }
+
+  @Override
+  public void postDeleteNamespace(ObserverContext<MasterCoprocessorEnvironment> ctx,
+      String namespace) throws IOException {
+    if (checkInitialized()) {
+      try (Table aclTable =
+          ctx.getEnvironment().getConnection().getTable(PermissionStorage.ACL_TABLE_NAME)) {
+        SnapshotScannerHDFSAclStorage.deleteNamespaceHdfsAcl(aclTable, namespace);
+      }
+      Path tmpNsDir = pathHelper.getTmpNsDir(namespace);
```

Comment:
Why we will remain a tmp/ns directory here ?   In my mind,   after the namespace deleted,     all the mob/ns, data/ns, tmp/ns should all be removed ....  Just confuse here why handle the tmp/ns separately..